deduction_rules.py

```diff
@@ -167,4 +167,162 @@ def line_square(sudoku):
                     r,c=local_to_global(sec,i,j)
                     if c!=col:
                         made_deduction|=sudoku.ban(r,c,val+1,"col_square",reason)
+    return made_deduction
+
+def naked_trios(sudoku):
+    """RULE: if v,w,x can be written only to three cells in this row/col/sec, then remove it from the other cells of the row/col/sec."""
+
+    def search_for_naked_trios(elems):
+        """Returns a list of 3-tuples, where every tuple contains the indeces of the identical inputs. The cases:
+        (123) (123) (123) - {3/3/3} (in terms of candidates per cell)
+        (123) (123) (12) - {3/3/2} (or some combination thereof)
+        (123) (12) (23) - {3/2/2/}
+        (12) (23) (13) - {2/2/2}
+        """
+        l = []
+        for comb in combinations(range(9),3):
+            cell0 = set(elems[comb[0]])
+            cell1 = set(elems[comb[1]])
+            cell2 = set(elems[comb[2]])
+            for cell0, cell1, cell2 in permutations((cell0, cell1, cell2)):
+                if len(cell0) == 3 and cell0 == cell1 and cell1 == cell2:
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 3 and cell0 == cell1 and len(cell2) == 2 and cell2.issubset(cell1):
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 3 and len(cell1) == 2 and len(cell2) == 2 and \
+                    cell1.issubset(cell0) and cell2.issubset(cell0) and len(cell1.intersection(cell2)) == 1:
+                    l.append(tuple(sorted((comb[0],comb[1],comb[2]))))
+                    break
+                elif len(cell0) == 2 and len(cell1) == 2 and len(cell1) == 2 and \
+                    len(set.intersection(cell0,cell1)) == 2 and len(set.intersection(cell1,cell2)) == 2 and \
+                    len(set.intersection(cell0,cell2)) == 2 and len(set.intersection(cell0,cell1,cell2)) == 1:
```

Comment:
The `set.intersection` function can be replaced by the `&` operator: `set_a & set_b == set.intersection(a, b)`. This makes the code more concise.

Reply:
fixed